tracing-subscriber/src/filter/targets.rs

```diff
@@ -266,6 +266,52 @@ impl Targets {
         self
     }
 
+    /// Returns an iterator over the [target]-[`LevelFilter`] pairs in this filter.
+    ///
+    /// The order of iteration is undefined.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tracing_subscriber::filter::{Targets, LevelFilter};
+    /// use tracing_core::Level;
+    ///
+    /// let filter = Targets::new()
+    ///     .with_target("my_crate", Level::INFO)
+    ///     .with_target("my_crate::interesting_module", Level::DEBUG);
+    ///
+    /// let mut targets: Vec<_> = filter.targets().collect();
+    /// targets.sort();
+    ///
+    /// assert_eq!(targets, vec![
+    ///     ("my_crate", LevelFilter::INFO),
+    ///     ("my_crate::interesting_module", LevelFilter::DEBUG),
+    /// ]);
+    /// ```
+    ///
+    /// This can be used to combine two [`Targets`]:
+    ///
+    /// ```
+    /// use tracing_subscriber::filter::Targets;
+    /// use tracing_core::Level;
+    ///
+    /// let mut filter = Targets::new().with_target("my_crate", Level::INFO);
+    /// let overrides = Targets::new().with_target("my_crate::interesting_module", Level::DEBUG);
+    ///
+    /// filter.extend(overrides.targets());
+    /// # drop(filter);
+    /// ```
+    ///
+    /// [target]: tracing_core::Metadata::target
+    pub fn targets(&self) -> impl Iterator<Item = (&str, LevelFilter)> + '_ {
```

Comment:
WDYT about also adding an `impl IntoIterator for &'_ Targets` that calls this? it seems like it might be nice to be able to write
```rust
let targets = // ...
for (target, level) in &targets {
     // ...
}
```

however, the downside is that this would probably require returning a named type rather than an `impl Iterator`. the best way to do that would probably be to change `DirectiveSet::directives` to return a `std::slice::Iter` rather than `impl Iterator`, here
https://github.com/tokio-rs/tracing/blob/b917104630efa0191a974fdbdf3a156bd78a5c9b/tracing-subscriber/src/filter/directive.rs#L65
and then adding an `Iter` type in this module like
```rust
pub struct Iter<'a>(
    std::iter::FilterMap<
        std::slice::Iter<'a, StaticDirective>,
        fn(&'a StaticDirective) -> Option<(&'a str, LevelFilter)>,
    >
);
```
and an `impl Iterator` for it.

this is a big pile of boilerplate code, so I'm not _totally_ convinced that it's worth the effort just to be able to add an `IntoIterator` impl so people can `for` loop over an `&Targets`...

however, a named iterator type might also be nice if users _also_ want to be able to return a named iterator type when composing the returned iterator from `Targets` with other iterator adapters. i'm not sure if there's enough of a use-case for that to matter, though. and, I think it's probably okay to go back and change this in the future if anyone *does* need a named return type: replacing an `impl Trait` return value with a named type is *probably* not a breaking change as long as the new return type implements the same trait (and has the correct lifetime etc).

what do you think?

Reply:
Not at all opposed to adding this. My two reservations with `IntoIterator`, which may be moot, were:

1. I was influenced by the implementation to consider whether the default `LevelFilter` should be yielded by an `IntoIterator` implementation, in order to properly iterate all the contained directives, but I don't think that would be very useful (but perhaps a getter for the default level would be useful?).

2. To have the `tracing_subscriber::filter::targets::Iter` type, the `tracing_subscriber::filter::targets` module would have to be made public. That would change the canonical location `Targets` itself. Not a breaking change since the re-export would remain, but something that felt 'too big' for a speculative PR 😅 

As I say, if these are both non-issues I'm happy to add this as described.

Reply:
Thinking more on this and your additional thoughts above, if we did add `IntoIterator` for `&Targets`, I'd be inclined to add `IntoIterator` for `Targets` as well, yielding `(String, LevelFIlter)`. That way we would also tick the box for having a way to merge `Targets` without cloning.

Reply:
Yeah, that could be nice. But, I'm on the fence about whether or not this is really worth the effort. It's up to you --- I'd happily merge this change as-is.

I'm not concerned about the canonical module path change (as long as the re-export doesn't go away), but I agree that we should probably add a way to access the default level if we're making the targets iterable. It would be fine to do that separately, though.

Reply:
No worries – I'll put in a commit with both, and we can see how it looks.